src/render/gif.rs

```diff
@@ -0,0 +1,278 @@
+use std::{sync::atomic::{AtomicUsize, Ordering}, fs::File, rc::Rc, cell::RefCell, time::{Instant, Duration}};
+
+use apex_hardware::FrameBuffer;
+
+use embedded_graphics::{prelude::{Point}, image::{ImageRaw, Image}, pixelcolor::BinaryColor, Drawable};
+use gif::{Frame};
+
+static GIF_MISSING: &[u8] = include_bytes!("./../../assets/gif_missing.gif");
+
+
+static DISPLAY_HEIGHT: i32 = 40;
+static DISPLAY_WIDTH: i32 = 128;
+
+pub struct Gif{
+    stop: Point,
+    origin: Point,
+    decoded_frames: Vec<Vec<u8>>,
+    current_frame: AtomicUsize,
+	delays: Vec<u16>,
+	time_frame_last_update: Rc<RefCell<Instant>>
+}
+
+
+impl Gif{
+
+	pub fn calculate_median_color_value(frame: &Frame, gif_height: i32, gif_width: i32) -> u8 {
+		//NOTE we're using the median to determine wether the pixel should be black or white 
+
+		let mut colors = (0..=255).into_iter().map(|_| 0).collect::<Vec<u32>>();
+	
+		//the u64 is just in case someone put a gif that's huge (in terms of resolution), it shouldn't break
+		let width= u64::from(frame.width);
+		let height= u64::from(frame.height);
+		
+		let num_pixels = gif_width as u32 * gif_height as u32;
+
+		let pixels = &frame.buffer;
+
+		for y in 0..gif_height{
+			//if y is outside of the gif width
+			if y as u64 >= height{
+				continue;
+			}
+
+			//if y is outside of the screen
+			if y >= DISPLAY_HEIGHT{
+				continue;
+			}
+			for x in 0..gif_width{
+				//if x is outside of the gif width
+				if x as u64 >= width{
+					continue;
+				}
+
+				//if x is outside of the screen
+				if x >= DISPLAY_WIDTH{
+					continue;
+				}
+
+				//calculating the index
+				let start:u64 = ((y as u64) * width + (x as u64))*4;
+
+				//getting the value of the pixels
+				let pixel_r = pixels.get(start as usize).unwrap_or(&0);
+				let pixel_g = pixels.get((start+1) as usize).unwrap_or(&0);
+				let pixel_b = pixels.get((start+2) as usize).unwrap_or(&0);
+				let pixel_a = pixels.get((start+3) as usize).unwrap_or(&0);
+
+				//the value is multiplied by the alpha of said pixel
+				//the more the pixel is transparent, the less the pixel has an importance
+				colors [(pixel_r/3 + pixel_g/3 + pixel_b/3) as usize] += u32::from(pixel_a / 255);
+			}
+		}
+	
+		let mut sum = 0;
+		for (color_value, count) in colors.iter().enumerate() {
+			sum += *count;
+	
+			if u32::from(sum) >= num_pixels / 2 {
+				if color_value == 0{
+					return 1;
+				}
+				return color_value as u8;
+			}
+		}
+	
+		1
+	}
+
+	pub fn read_frame(frame: &Frame, gif_height: i32, gif_width: i32) -> Vec<u8>{
+
+		let median_color = Self::calculate_median_color_value(frame, gif_height, gif_width);
+
+		let mut image = Vec::new();
+		let mut buf: u8 = 0;
+
+		//the u64 is just in case someone put a gif that's huge (in terms of resolution), it shouldn't break

```

Comment:
These can't overflow if you convert *from* u16 :p

Reply:
I tried (not sure it that was what you meant) to not convert them to u64 (I left the comment, will correct)